### PR TITLE
All changes belonging to one package are put in one folder under changes root folder.

### PR DESCRIPTION
**Pull request by qz2017**
This change is [Reviewable](https://reviewable.io/reviews/microsoft/web-build-tools/177)

### CODE REVIEW

File: rush/rush/src/utilities/ChangeFiles.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import * as fsx from 'fs-extra';
+import * as path from 'path';
+import { EOL } from 'os';
+import {
+  Utilities,
+  IChangeInfo
+} from '@microsoft/rush-lib';
+
+export default class ChangeFiles {
```
Comment (pgonzal) [inline rush/rush/src/utilities/ChangeFiles.ts]: >ChangeFiles [](start = 21, length = 11)

Add JSDoc explaining what this does, and which command(s) use it

File: rush/rush-lib/src/utilities/Utilities.ts [hunk 2]
```diff
@@ -142,6 +143,16 @@ export default class Utilities {
   }
 
   /**
+   * return all files in the folder recursively through all sub-directories.
```
Comment (pgonzal) [inline rush/rush-lib/src/utilities/Utilities.ts]: > return all files in the folder recursively through all sub-directories. [](start = 4, length = 72)

How useful is this "reusable" utility?  Would it be simpler to use the glob.sync() API instead?

File: rush/rush-lib/src/utilities/Utilities.ts [hunk 3]
```diff
@@ -351,4 +362,20 @@ export default class Utilities {
...
+        const fullSubPath: string = path.join(folder, file);
+        const lstat: fsx.Stats = fsx.lstatSync(fullSubPath);
+        if (lstat.isDirectory()) {
+          Utilities._getFiles(fullSubPath, files);
+        } else if (lstat.isFile()) {
+          files.push(fullSubPath);
+        }
+      });
+    } catch (e) {
+      /* no-op when empty folder */
```
Comment (pgonzal) [inline rush/rush-lib/src/utilities/Utilities.ts]: >/* no-op when empty folder */ [](start = 6, length = 29)

Avoid throwing/catching "unexceptional exceptions"

File: rush/rush/src/actions/ChangeAction.ts [hunk 4]
```diff
@@ -355,14 +357,31 @@ export default class ChangeAction extends CommandLineAction {
...
 
+  private _generateChangeFilePath(packageName: string): string {
+    let branch: string = undefined;
+    try {
+      branch = gitInfo().branch;
+    } catch (error) {
+      console.log('Could not automatically detect git branch name, using timestamps instead.');
+    }
+
+    const filename: string = (branch ?
```
Comment (pgonzal) [inline rush/rush/src/actions/ChangeAction.ts]: >filename [](start = 10, length = 8)

Whenever you are pasting strings together to make a file/path, add a TypeScript comment showing an example of what it looks like.  Something like this:

```
    const filename: string = (branch ?
      // Example: "MyBranch_20160502.json"
      this._escapeFilename(`${branch}_${this._getTimestamp()}.json`)
      // Example: "20160502.json"
      : `${this._getTimestamp()}.json`);

    // Example: "./common/changes/@scope/package/MyBranch_20160502.json"
    const filepath: string = path.join(this._rushConfiguration.commonFolder,
      'changes',
      ...packageName.split('/'),
      filename);

```

This gives a huge improvement in readability.

**Comment by pgonzal** [comment]
:clock1:

File: rush/rush/src/actions/ChangeAction.ts (hunk 4 above)
Comment (qz2017) [inline rush/rush/src/actions/ChangeAction.ts]: Added example.

File: rush/rush/src/utilities/ChangeFiles.ts (hunk 1 above)
Comment (qz2017) [inline rush/rush/src/utilities/ChangeFiles.ts]: Added.

File: rush/rush-lib/src/utilities/Utilities.ts (hunk 2 above)
Comment (qz2017) [inline rush/rush-lib/src/utilities/Utilities.ts]: Removed and replaced.

File: rush/rush-lib/src/utilities/Utilities.ts (hunk 3 above)
Comment (qz2017) [inline rush/rush-lib/src/utilities/Utilities.ts]: code removed.

**Pull request merged**